Mage.Sets/src/mage/cards/z/ZaxaraTheExemplary.java

```diff
@@ -119,7 +120,8 @@ public boolean apply(Game game, Ability source) {
             // create token
             if (needObject instanceof Spell) {
                 Spell spell = (Spell) needObject;
-                int xValue = spell.getSpellAbility().getManaCostsToPay().getX();
+
+                int xValue = CardUtil.getSourceCostsTag(game, spell.getSpellAbility(), "X", 0);
```

Comment:
1. What about containsX? Why not replaced?

![IMG_0131](https://github.com/magefree/mage/assets/8344157/705a4c7e-aa30-4940-9654-8ef1b834a7d8)

2. There are many other methods with X value. Explain: a. is it same/compatible with coststag? b. Is it require refactor/replace too?

![IMG_0132](https://github.com/magefree/mage/assets/8344157/0e799374-a055-4691-a294-4cccf49d973b)

Reply:
`containsX` should probably be removed, yeah. I'll do that in a bit, drafting it for now until I do so.

`setX` is still necessary: the remaining uses of `getX` need it to function correctly, and there are some card implementations that access the X value in other ways that this PR doesn't cover (such as implementing their own search for the cost). I would like to convert those eventually, but I figured it would be better to adjust these relatively easy ones now with a find/replace and get to the others later. All `Costs.getX()` calls used basically the same pattern, which made it easy and safe to do a mass replacement like this.

Reply:
> and there are some card implementations that access the X value in other ways that this PR doesn't cover

What you mean by "other ways"? Write examples here. If tags return wrong values then something wrong and must be improved.

Reply:
Tags would provide the right values, just there's sometimes complex logic involved. Example: `ExplosiveSingularityCostReductionEffect`
![image](https://github.com/magefree/mage/assets/8648826/a0f220e6-c33f-4493-8399-1f942b75aff5)
"real cast" here could use the tag system, but I couldn't easily get ones like this with a simple find/replace regex and there's a bunch of cards that use a similar style of code where they do the loop through the costs to find the variable one.

Reply:
It's uses `getCosts`, not `getX` here. I still don't see examples/reason why you keep that method (`getX`) in current PR

Reply:
That was for `setX`, which sets the value looked up here.

Looking into it further I think I can actually remove the last few `getX` calls, though I will have to access the ability's tag map directly instead of going through the standard CardUtil function (which requires the Game object). I'll investigate further before un-drafting.

Specifically: `SpellAbility.getConvertedXManaCost(Card card)` which is by `Spell.getManaValue()`. This is a problem because the `MageObject.getManaValue()` interface function does not take the `Game` object, but the spell's mana value needs to depend on the value chosen for X. However the information is available, so it should be fine.

Reply:
Also it should be noted that the logic above with `getCosts` is the same logic that `getX` uses, so even though it isn't calling the function it's still basically equivalent.

Reply:
Are you planning to remove `VariableCostImpl` usage or do something with `VariableManaCost`?

![shot_240404_061642](https://github.com/magefree/mage/assets/8344157/80a35df5-a19a-404d-9a32-10494b3d98df)

Reply:
Eventually I would like to remove `VariableCost.getAmount()`, since I think all uses of it can be replaced with tags. Right now it's duplicated between the tag system and inside the cost. This is not likely to be done in the near future, though.

Reply:
On further investigation, `containsX` is still necessary. It is used to check if a spell has {X} *in its mana cost*, whereas the tags are for any spell with X involved in its casting, such as [[Harvest Pyre]]. However, that does make me think that `CardView`'s check for X should probably be changed to include those sorts of spells.